src/driver-logitech-g300.c

```diff
@@ -0,0 +1,492 @@
+		if (mapping->raw == data)
+			return mapping->type;
+	}
+
+	return RATBAG_BUTTON_TYPE_UNKNOWN;
+}
+
+struct logitech_g300_button_mapping {
+	uint8_t raw;
+	struct ratbag_button_action action;
+};
+
+static struct logitech_g300_button_mapping logitech_g300_button_mapping[] = {
+	/* 0x00 is either key or unassigned. Must be handled separatly  */
+	{ 0x01, BUTTON_ACTION_BUTTON(RATBAG_BUTTON_TYPE_LEFT) },
+	{ 0x02, BUTTON_ACTION_BUTTON(RATBAG_BUTTON_TYPE_RIGHT) },
+	{ 0x03, BUTTON_ACTION_BUTTON(RATBAG_BUTTON_TYPE_MIDDLE) },
+	{ 0x04, BUTTON_ACTION_BUTTON(RATBAG_BUTTON_TYPE_THUMB) },
+	{ 0x05, BUTTON_ACTION_BUTTON(RATBAG_BUTTON_TYPE_THUMB2) },
+	{ 0x0A, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_UP) },
+	{ 0x0B, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_DOWN) },
+	{ 0x0C, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_CYCLE_UP) },
+	{ 0x0D, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_CYCLE_UP) },
+	{ 0x0E, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_ALTERNATE) },
+	{ 0x0F, BUTTON_ACTION_SPECIAL(RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_DEFAULT) },
+};
+
+static const struct ratbag_button_action*
+logitech_g300_raw_to_button_action(uint8_t data)
+{
+	struct logitech_g300_button_mapping *mapping;
+
+	ARRAY_FOR_EACH(logitech_g300_button_mapping, mapping) {
+		if (mapping->raw == data)
+			return &mapping->action;
+	}
+
+	return NULL;
+}
+
+static uint8_t
+logitech_g300_button_action_to_raw(const struct ratbag_button_action *action)
+{
+	struct logitech_g300_button_mapping *mapping;
+
+	ARRAY_FOR_EACH(logitech_g300_button_mapping, mapping) {
+		if (ratbag_button_action_match(&mapping->action, action))
+			return mapping->raw;
+	}
+
+	return 0;
+}
+
+struct logitech_g300_frequency_mapping {
+	uint8_t raw;
+	unsigned int frequency;
+};
+
+static struct logitech_g300_frequency_mapping logitech_g300_frequency_mapping[] = {
+	{ 0, 1000 },
+	{ 1, 125 },
+	{ 2, 250 },
+	{ 3, 500 },
+};
+
+static unsigned int
+logitech_g300_raw_to_frequency(uint8_t data)
+{
+	struct logitech_g300_frequency_mapping *mapping;
+
+	ARRAY_FOR_EACH(logitech_g300_frequency_mapping, mapping) {
+		if (mapping->raw == data)
+			return mapping->frequency;
+	}
+
+	return 0;
+}
+
+static uint8_t
+logitech_g300_frequency_to_raw(unsigned int frequency)
+{
+	struct logitech_g300_frequency_mapping *mapping;
+
+	ARRAY_FOR_EACH(logitech_g300_frequency_mapping, mapping) {
+		if (mapping->frequency == frequency)
+			return mapping->raw;
+	}
+
+	return 0;
+}
+
+static int
+logitech_g300_current_profile(struct ratbag_device *device)
+{
+	uint8_t buf[LOGITECH_G300_REPORT_SIZE_ACTIVE_PROFILE];
+	int ret;
+
+	ret = ratbag_hidraw_raw_request(device, LOGITECH_G300_REPORT_ID_GET_ACTIVE_PROFILE, buf,
+			sizeof(buf), HID_FEATURE_REPORT, HID_REQ_GET_REPORT);
+
+	if (ret < 0)
+		return ret;
+
+	if (ret != sizeof(buf))
+		return -EIO;
+
+	switch (buf[1]) {
+	case 0x07: return 0;
+	case 0x02: return 1;
+	case 0x04: return 2;
```

Comment:
You are right. It is the color. Which is strange since two profiles can use the same color.